vxsandbox/resources/tests/test_utils.py

```diff
@@ -1 +1,288 @@
+        self.assertEqual(resources.app_worker, dummy_worker)
+        self.assertEqual(resources.config, {"conf": "foo"})
+        self.assertEqual(resources.resources, {})
+
+    def test_add_resource(self):
+        resources = self.mk_resources()
+        tst = RecordingResource("tst", resources.app_worker, {})
+        self.assertEqual(resources.resources, {})
+        resources.add_resource("tst", tst)
+        self.assertEqual(resources.resources, {"tst": tst})
+        self.assertEqual(tst.setup_calls, 0)
+        self.assertEqual(tst.teardown_calls, 0)
+
+    def test_validate_config(self):
+        resources = self.mk_resources(config={
+            'tst1': {
+                'cls': self.class_name(RecordingResource),
+            },
+            'tst2': {
+                'cls': self.class_name(RecordingResource),
+                'extra': 'more config'
+            },
+        })
+        self.assertEqual(resources.resources, {})
+        resources.validate_config()
+        self.assert_recording_resources(resources, {
+            'tst1': {},
+            'tst2': {'config': {'extra': 'more config'}}
+        })
+
+    def test_setup_resources(self):
+        resources = self.mk_resources(config={
+            'tst1': {'cls': self.class_name(RecordingResource)},
+            'tst2': {'cls': self.class_name(RecordingResource)},
+        })
+        resources.validate_config()
+        self.assert_recording_resources(resources, {
+            'tst1': {'setup_calls': 0},
+            'tst2': {'setup_calls': 0},
+        })
+        resources.setup_resources()
+        self.assert_recording_resources(resources, {
+            'tst1': {'setup_calls': 1},
+            'tst2': {'setup_calls': 1},
+
+        })
+
+    def test_teardown_resources(self):
+        resources = self.mk_resources(config={
+            'tst1': {'cls': self.class_name(RecordingResource)},
+            'tst2': {'cls': self.class_name(RecordingResource)},
+        })
+        resources.validate_config()
+        self.assert_recording_resources(resources, {
+            'tst1': {'teardown_calls': 0},
+            'tst2': {'teardown_calls': 0},
+        })
+        resources.teardown_resources()
+        self.assert_recording_resources(resources, {
+            'tst1': {'teardown_calls': 1},
+            'tst2': {'teardown_calls': 1},
+        })
+
+
+class CommandHandler(object):
+    def __init__(self, name, test_case, result=None):
+        self._name = name
+        self._test_case = test_case
+        self._calls = []
+        self._result = result or object()
+
+    def assert_called(self, api, cmd, result_d):
+        self._test_case.assertEqual(self._calls, [(api, cmd)])
+        if isinstance(self._result, Deferred):
+            self._test_case.assertEqual(result_d, self._result)
+        else:
+            self._test_case.assertTrue(result_d.called)
+            self._test_case.assertEqual(result_d.result, self._result)
```

Comment:
This is testing the handler dispatcher's use of ``maybeDeferred``.
``result_d`` is (or should always be) a deferred, but what is returned by
the handler (``self._result``) might be either a deferred (in which case it
should be identical to ``result_d``) or not (in which case ``result_d``
should already have fired with ``self._result``).